src/python/pants/backend/docker/rules_context.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+from dataclasses import dataclass
+from itertools import chain
+from typing import Tuple
+
+from pants.backend.docker.target_types import DockerImageSources
+from pants.core.goals.package import BuiltPackage, PackageFieldSet
+from pants.core.target_types import FilesSources, ResourcesSources
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Address
+from pants.engine.fs import AddPrefix, Digest, MergeDigests
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    Dependencies,
+    DependenciesRequest,
+    FieldSetsPerTarget,
+    FieldSetsPerTargetRequest,
+    Sources,
+    Target,
+    Targets,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class DockerBuildContext:
+    digest: Digest
+
+
+@dataclass(frozen=True)
+class DockerBuildContextRequest:
+    address: Address
+    context_root: str
+    targets: Tuple[Target, ...]
+
+
+@rule
+async def create_docker_build_context(request: DockerBuildContextRequest) -> DockerBuildContext:
+    # Get all sources, going into the build context.
+    sources_request = Get(
+        SourceFiles,
+        SourceFilesRequest(
+            sources_fields=[t.get(Sources) for t in request.targets],
+            for_sources_types=(DockerImageSources, FilesSources, ResourcesSources),
+        ),
+    )
+
+    # Get all dependent targets.
+    dependencies_targets = await MultiGet(
```

Comment:
Since we're already operating on the transitive closure, not sure why we need to get dependencies? This would be clearer if the expansion to transitive closure was in this rule.

Reply:
Instead you can do something like:

```
pkg_field_sets_per_tgt = await Get(
    FieldSetsPerTarget, FieldSetsPerTargetRequest(PackageFieldSet, request.targets)
)
pkgs = await MultiGet(
    Get(BuiltPackage, PackageFieldSet, field_set) for field_set in pkg_field_sets_per_tgt.field_sets
)
```

I think. @Eric-Arellano does that look right?

Reply:
Yeah, as long as I make sure to filter out the current target from that list. Which the `DependenciesRequest` seemed to do for me?!?...

```
TARGETS: 
testprojects/src/python/docker/Dockerfile:test-example
testprojects/src/python/hello/main
testprojects/src/python/hello/main/main.py:lib
testprojects/src/python/hello/greet/greet.py
3rdparty/python:ansicolors
3rdparty/python:requirements.txt

09:16:47.49 [DEBUG] Get packages for Docker image from:
testprojects/src/python/hello/main
testprojects/src/python/hello/main/main.py:lib
testprojects/src/python/hello/greet/greet.py
3rdparty/python:ansicolors
3rdparty/python:requirements.txt
```

```python
    logger.debug(f"\n\nTARGETS: {', '.join(str(t.address) for t in request.targets)}\n\n")

    # Get all dependent targets.
    dependencies_targets = await MultiGet(
        Get(Targets, DependenciesRequest(t.get(Dependencies))) for t in request.targets
    )

    embedded_pkg_tgts = tuple(chain(*dependencies_targets))
    logger.debug(
        f"""Get packages for Docker image from: {", ".join(str(t.address) for t in embedded_pkg_tgts)}"""
    )
```

Reply:
I sometimes forget that the v2 engine is so powerful, the naive thing just works by design... In this case, packaging dependencies will recurse as needed... :)